ls1intum/Artemis
# Export metrics in prometheus format

### Checklist

### Motivation and Context
Spring collects several metrics, including cache performance, http performance and system health.
These metrics are displayed in the `Metrics` Tab of both Artemis and the Jhipster Registry.
As we now increasingly use external monitoring systems, we also want to publish insights into the Artemis application to these systems.
This PR adds additional metrics (Health status and User count) to the prometheus endpoint and adds an configuration option to specify the ip address of the monitoring system, that should be allowed to access these metrics.

### Description
Due to Spring not exporting the Health status for the individual indicators (See https://github.com/micrometer-metrics/micrometer/issues/416 and https://github.com/spring-projects/spring-boot/issues/14087), I've added these metrics manually using a micrometer `Gauge`.
The current user count and the status of the system will then be published, alongside the 'normal' metrics like http request time, cpu usage, etc and can be collected by Prometheus.

To allow only the monitoring system to access these metrics, a new configuration option `spring.prometheus.monitoringIp` is added (defaulting to `127.0.0.1`), that must equal the IP address of the monitoring system (that is used to access the Prometheus endpoint).


### Steps for Testing (locally)
1. Change the config value `management.metrics.export.prometheus.enabled` to `true`.
2. If necessary, adjust the value of `spring.prometheus.monitoringIp` to match the IP address used to access Artemis.
3. Navigate to `ARTEMIS_IP:8080/management/prometheus` (where `ARTEMIS_IP` is `127.0.0.1` when using the same device) and verify that something like this is shown:
![image](https://user-images.githubusercontent.com/5084100/87094389-1a686980-c23f-11ea-95ee-9b7cd8e16a04.png)

4. Change to config value of `spring.prometheus.monitoringIp` to a different ip address & reboot Artemis.
5. Verify that the metrics can not be accessed anymore:
![image](https://user-images.githubusercontent.com/5084100/87094489-4126a000-c23f-11ea-9d7f-c4b78980c769.png)
